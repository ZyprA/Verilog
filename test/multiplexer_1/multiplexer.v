module multiplexer(a, b, c, f, s);
    input a, b, c;
    input [1:0] f;
    output s;

    function mux;
        input [1:0] s1;
        input d0, d1, d2;

        case (s1)
            2'b00 : mux = d0;
            2'b01 : mux = d1;
            2'b10 : mux = d2;
            default : mux = 1'bx;
        endcase
    endfunction

    assign s = mux(f,a,b,c);

endmodule


`timescale 1ns / 1ns

module multiplexer_sim;
    reg a, b, c;
    reg [1:0] f;
    wire s;

    multiplexer uut(a, b, c, f, s);

    initial begin
        $dumpfile("multiplexer.vcd");
        $dumpvars(0, multiplexer_sim);
    end

    initial begin
        a = 0; b = 0; c = 1; f = 0;
        #100 f = 0;
        #100 f = 2;
        #100 f = 1;
        #100 f = 3;
        #100 $finish;
    end

endmodule