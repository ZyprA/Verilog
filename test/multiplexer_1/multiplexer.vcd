$date
	Sun Feb  2 15:35:37 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1ns
$end
$scope module multiplexer_sim $end
$var wire 1 ! s $end
$var reg 1 " a $end
$var reg 1 # b $end
$var reg 1 $ c $end
$var reg 2 % f [1:0] $end
$scope module uut $end
$var wire 1 " a $end
$var wire 1 # b $end
$var wire 1 $ c $end
$var wire 2 & f [1:0] $end
$var wire 1 ! s $end
$scope function mux $end
$var reg 1 ' d0 $end
$var reg 1 ( d1 $end
$var reg 1 ) d2 $end
$var reg 2 * s1 [1:0] $end
$upscope $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
b0 *
1)
0(
0'
b0 &
b0 %
1$
0#
0"
0!
$end
#200
1!
b10 *
b10 %
b10 &
#300
0!
b1 *
b1 %
b1 &
#400
x!
b11 *
b11 %
b11 &
#500
