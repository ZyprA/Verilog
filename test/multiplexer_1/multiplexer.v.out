#! /opt/homebrew/Cellar/icarus-verilog/12.0/bin/vvp
:ivl_version "12.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision - 9;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/va_math.vpi";
S_0x13e60e740 .scope module, "multiplexer_sim" "multiplexer_sim" 2 25;
 .timescale -9 -9;
v0x13e61fc10_0 .var "a", 0 0;
v0x13e61fcc0_0 .var "b", 0 0;
v0x13e61fd50_0 .var "c", 0 0;
v0x13e61fe00_0 .var "f", 1 0;
v0x13e61fe90_0 .net "s", 0 0, L_0x13e61ff60;  1 drivers
S_0x13e609930 .scope module, "uut" "multiplexer" 2 30, 2 1 0, S_0x13e60e740;
 .timescale 0 0;
    .port_info 0 /INPUT 1 "a";
    .port_info 1 /INPUT 1 "b";
    .port_info 2 /INPUT 1 "c";
    .port_info 3 /INPUT 2 "f";
    .port_info 4 /OUTPUT 1 "s";
v0x13e61f820_0 .net "a", 0 0, v0x13e61fc10_0;  1 drivers
v0x13e61f8d0_0 .net "b", 0 0, v0x13e61fcc0_0;  1 drivers
v0x13e61f970_0 .net "c", 0 0, v0x13e61fd50_0;  1 drivers
v0x13e61fa00_0 .net "f", 1 0, v0x13e61fe00_0;  1 drivers
v0x13e61fab0_0 .net "s", 0 0, L_0x13e61ff60;  alias, 1 drivers
L_0x13e61ff60 .ufunc/vec4 TD_multiplexer_sim.uut.mux, 1, v0x13e61fe00_0, v0x13e61fc10_0, v0x13e61fcc0_0, v0x13e61fd50_0 (v0x13e61f730_0, v0x13e607530_0, v0x13e61f560_0, v0x13e61f600_0) S_0x13e609aa0;
S_0x13e609aa0 .scope function.vec4.s1, "mux" "mux" 2 6, 2 6 0, S_0x13e609930;
 .timescale 0 0;
v0x13e607530_0 .var "d0", 0 0;
v0x13e61f560_0 .var "d1", 0 0;
v0x13e61f600_0 .var "d2", 0 0;
; Variable mux is vec4 return value of scope S_0x13e609aa0
v0x13e61f730_0 .var "s1", 1 0;
TD_multiplexer_sim.uut.mux ;
    %load/vec4 v0x13e61f730_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 2;
    %cmp/u;
    %jmp/1 T_0.0, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 2;
    %cmp/u;
    %jmp/1 T_0.1, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 2;
    %cmp/u;
    %jmp/1 T_0.2, 6;
    %pushi/vec4 1, 1, 1;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
    %jmp T_0.4;
T_0.0 ;
    %load/vec4 v0x13e607530_0;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
    %jmp T_0.4;
T_0.1 ;
    %load/vec4 v0x13e61f560_0;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
    %jmp T_0.4;
T_0.2 ;
    %load/vec4 v0x13e61f600_0;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
    %jmp T_0.4;
T_0.4 ;
    %pop/vec4 1;
    %end;
    .scope S_0x13e60e740;
T_1 ;
    %vpi_call 2 33 "$dumpfile", "multiplexer.vcd" {0 0 0};
    %vpi_call 2 34 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x13e60e740 {0 0 0};
    %end;
    .thread T_1;
    .scope S_0x13e60e740;
T_2 ;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13e61fc10_0, 0, 1;
    %pushi/vec4 0, 0, 1;
    %store/vec4 v0x13e61fcc0_0, 0, 1;
    %pushi/vec4 1, 0, 1;
    %store/vec4 v0x13e61fd50_0, 0, 1;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v0x13e61fe00_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v0x13e61fe00_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 2, 0, 2;
    %store/vec4 v0x13e61fe00_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 1, 0, 2;
    %store/vec4 v0x13e61fe00_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 3, 0, 2;
    %store/vec4 v0x13e61fe00_0, 0, 2;
    %delay 100, 0;
    %vpi_call 2 43 "$finish" {0 0 0};
    %end;
    .thread T_2;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "multiplexer.v";
